piet/src/render_context.rs

```diff
@@ -73,6 +73,13 @@ pub trait RenderContext {
     /// The associated type of an image.
     type Image;
 
+    /// Report an internal error.
+    ///
+    /// Drawing operations may cause internal errors, which may also occur
+    /// asynchronously after the drawing command was issued. This method reports
+    /// any such error that has been detected.
+    fn status(&mut self) -> Result<(), Error>;
```

Comment:
[Only when the function itself also returns Option](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0ddabfd042406131562a48b771c90000). I think `Result` is the most idiomatic way to indicate that it might be an error.

Reply:
I think that `Result<(), Error>` is more idiomatic, but the `status` method itself is definitely not. Mainly because you have to check it manually. Which is against the Rust philosophy, where all errors should be handled.

Reply:
@RazrFalcon Do you have a concrete suggestion what to do here? My idea is that a typical usage would simply be `rc.status()?` for a method that returned a Result. I think other choices would be much more verbose, requiring explicit match (I can't easily find a conversion like [`Result::ok`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) but in the other direction).

Reply:
Another (arguably very overengineered option) is to have what is now the `RenderContext` only have one method `perform_draw` which takes a lambda that gives you access to the actual render context and all the drawing functions. This `perform_draw` function could then return a result that the caller would be forced to handle.

Reply:
That sounds too heavy to me, and I can see it making borrowing more difficult as it constrains control flow.

Also keep in mind that `finish` returns a `Result`. I think `status` won't get used much in practice; the main value I see in it is to help with debugging, to be able to retrieve errors reasonably early after they're created, rather than having to wait til the end of drawing.

Reply:
First of all, the `status()` method should explain what kind of errors may occur.  But, `ErrorKind` suggests, that anything can happen. So we can't really handle it, and so we don't really care...

Reply:
By the way, why do we need the `finish()` method? I thought that stuff will appear on the canvas immediately.

Reply:
Re why `finish()` is needed: the implementation of clipping in `Direct2D` is not based on immediately drawing clipped pixels to the final canvas, but rather drawing to an intermediate buffer and then compositing that buffer to the final surface with the clip region as a mask. If the clip is not enclosed in `save`/`restore` then it's done at `finish` time. (I think implementations can be more sophisticated but the model allows this).

Reply:
Yes, it's complicated. That's why I didn't done this in the first place.

`cairo` have the same behavior, kinda, but Qt is always immediate, which makes it much easier to use.